src/probability_theory/martingale.lean

```diff
@@ -83,11 +112,149 @@ begin
   rw [pi.smul_apply, hx, pi.smul_apply, pi.smul_apply],
 end
 
+lemma supermartingale [preorder E] (hf : martingale f ℱ μ) : supermartingale f ℱ μ :=
+⟨hf.1, λ i j hij, (hf.2 i j hij).le, λ i, hf.integrable i⟩
+
+lemma submartingale [preorder E] (hf : martingale f ℱ μ) : submartingale f ℱ μ :=
+⟨hf.1, λ i j hij, (hf.2 i j hij).symm.le, λ i, hf.integrable i⟩
+
 end martingale
 
+lemma martingale_iff [partial_order E] : martingale f ℱ μ ↔
+  supermartingale f ℱ μ ∧ submartingale f ℱ μ :=
+⟨λ hf, ⟨hf.supermartingale, hf.submartingale⟩,
+ λ ⟨hf₁, hf₂⟩, ⟨hf₁.1, λ i j hij, (hf₁.2.1 i j hij).antisymm (hf₂.2.1 i j hij)⟩⟩
+
 lemma martingale_condexp (f : α → E) (ℱ : filtration ι m0) (μ : measure α)
   [sigma_finite_filtration μ ℱ] :
   martingale (λ i, μ[f | ℱ i, ℱ.le i]) ℱ μ :=
 ⟨λ i, measurable_condexp, λ i j hij, condexp_condexp_of_le (ℱ.mono hij) _⟩
 
+namespace supermartingale
+
+lemma adapted [has_le E] (hf : supermartingale f ℱ μ) : adapted ℱ f := hf.1
+
+lemma measurable [has_le E] (hf : supermartingale f ℱ μ) (i : ι) : measurable[ℱ i] (f i) :=
+hf.adapted i
+
+lemma integrable [has_le E] (hf : supermartingale f ℱ μ) (i : ι) : integrable (f i) μ := hf.2.2 i
+
+lemma condexp_ae_eq [has_le E] (hf : supermartingale f ℱ μ) {i j : ι} (hij : i ≤ j) :
+  μ[f j | ℱ i, ℱ.le i] ≤ᵐ[μ] f i :=
+hf.2.1 i j hij
+
+lemma set_integral_le {f : ι → α → ℝ} (hf : supermartingale f ℱ μ)
+  {i j : ι} (hij : i ≤ j) {s : set α} (hs : measurable_set[ℱ i] s) :
+  ∫ x in s, f j x ∂μ ≤ ∫ x in s, f i x ∂μ :=
+begin
+  rw ← @set_integral_condexp _ _ _ _ _ _ _ _ (ℱ i) m0 _ (ℱ.le i) _ _ _ (hf.integrable j) hs,
```

Comment:
could we make en argument explicit in `set_integral_condexp` and avoid that `@`?

Reply:
I've added `set_integral_condexp'` with the explicit argument for now. Do you think it is worth it to change `set_integral_condexp` directly? The argument is fine being implicit in the case there is only 2 sigma algebras.

Reply:
I think we should change `set_integral_condexp` directly. If we work with conditional expectations, we are likely to have several sigma algebras.